src/v2/components/TextInput.jsx

```diff
@@ -0,0 +1,44 @@
+// -----------------------------------------------------------------------------
+// Copyright 2022 Open Climate Tech Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// -----------------------------------------------------------------------------
+
+import React from 'react'
+
+import concatClassNames from '../modules/concatClassNames.mjs'
+
+/**
+ * @param {Object} props
+ * @param {string} props.className - A list of CSS class names.
+ * @param {string=} props.label - Optional label for the input field.
+ * @param {string=} props.placeholder - Optional placeholder for the input field
+ *     also used as accessible label if `label` isn’t specified.
+ *
+ * @returns {React.Element}
+ */
+export default function TextInput(props) {
+  const {className, disabled, label, placeholder, ...otherProps} = props
+  const classNames =
+    concatClassNames(className, 'c7e-text-input', disabled && 'c7e-text-input--disabled')
+
+  return 0,
+  <label>
+    { label
+      ? <span className="c7e-text-input--label">{label}</span>
+      : placeholder &&
+        <span className="c7e-text-input--label" style={{position: 'absolute', left: -99999}}>{placeholder}</span>
+    }
+    <input type="text" placeholder={placeholder} disabled={disabled} className={classNames} {...otherProps}/>
```

Comment:
I maybe should check whether `type` is specified in props and be explicit about the override. But it’s technically fine. `otherProps` is spread (i.e., `...`) after `type` in the JSX code, and the JSX is converted to straight JS like so:

```javascript
React.createElement(input, {type: 'text', placeholder, disabled, className ...otherProps})
```

So any property in `otherProps` will overwrite properties defined at the start of the `createElement()` `props` argument. Aside from the fact that it may lead to some confusion down the road (or in this review 😉), you can think of `type="text"` as setting the default value for `type` if it’s left unspecified.

It’s easy enough to be explicit, though:

```javascript
const {className, disabled, label, placeholder, type = 'text', ...otherProps} = props
```
...

```jsx
<input type={type} placeholder={placeholder} disabled={disabled} className={classNames} {...otherProps}/>
```

Reply:
Thinking about this some more, I’m not sure we want to anticipate every standard HTML attribute that might be overwritten—the list is just too long. The convention that I’ve followed in the past is to be disciplined about (a) placing the `{...otherProps}` spread last in JSX and (b) destructuring attributes that we explicitly _don’t_ want overwritten (e.g., `{type: ignoredType} = props`).